### PR TITLE
Connect Four: Backend

### PR DESCRIPTION
**Pull request by cdleveille**

### CODE REVIEW

File: src/controllers/gambling/index.ts [hunk 1]
```diff
@@ -177,4 +177,175 @@ export const gamblingRouter = async function (app: FastifyInstance) {
...
+				}
+				// update the existing challenge
+				return await connectFourGames.updateExistingChallenge(
+					existingChallenge,
+					mentionedUser,
+					wager
+				);
+			}
+			// issue a new challenge to the mentioned user
+			return await connectFourGames.createNewChallenge(user, mentionedUser, wager);
```
Comment (judegiordano) [inline src/controllers/gambling/index.ts:278]: swap this logic so it returns early first, the if outer statement is too big

File: src/controllers/gambling/index.ts [hunk 2]
```diff
@@ -177,4 +177,175 @@ export const gamblingRouter = async function (app: FastifyInstance) {
...
+			}
+		},
+		async (req) => {
+			const { server_id, user_id, move } = req.body;
+			await servers.assertExists({ server_id });
+
+			const user = await users.assertRead({ user_id, server_id });
+			const activeGame = await connectFourGames.getActiveGame(user);
+			if (!activeGame)
+				throw new BadRequestError("You do not have an active game of Connect Four!");
```
Comment (judegiordano) [inline src/controllers/gambling/index.ts]: could probably enclose this into its own method for asserting an active game

File: src/models/connect-four-games.ts [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+			server_id,
+			$or: [{ red_user_id: user_id }, { yellow_user_id: user_id }],
+			is_accepted: challenge ? false : true,
+			is_complete: false
+		});
+	}
+
+	/** Return user's open/unaccepted active challenge if it exists, else null */
+	public async getUnacceptedChallenge(user: IUser) {
+		return await this.getActiveGame(user, true);
```
Comment (judegiordano) [inline src/models/connect-four-games.ts]: dont await here, return the unresolved promise

File: src/models/connect-four-games.ts [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+				required: false,
+				default: false
+			}
+		});
+	}
+
+	/** Return user's active game or open/unaccepted challenge if it exists, else null */
+	public async getActiveGame(user: IUser, challenge?: boolean) {
+		const { user_id, server_id } = user;
+		return await super.read({
```
Comment (judegiordano) [inline src/models/connect-four-games.ts]: dont await here, return the unresolved promise

File: src/models/connect-four-games.ts [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+
+	/** Return true if user has an active game, else false */
+	public async hasActiveGame(user: IUser) {
+		if (await this.getActiveGame(user)) return true;
+		return false;
+	}
+
+	/** Challenge another player to a game */
+	public async createNewChallenge(challengerUser: IUser, challengedUser: IUser, wager: number) {
+		return await super.insertOne({
```
Comment (judegiordano) [inline src/models/connect-four-games.ts]: dont await here, return the unresolved promise

File: src/models/connect-four-games.ts [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+		challengedUser: IUser,
+		wager: number
+	) {
+		if (
+			existingChallenge.yellow_user_id === challengedUser.user_id &&
+			existingChallenge.wager === wager
+		)
+			return existingChallenge;
+		const { _id } = existingChallenge;
+		return await super.updateOne({ _id }, { yellow_user_id: challengedUser.user_id, wager });
```
Comment (judegiordano) [inline src/models/connect-four-games.ts]: dont await here, return the unresolved promise

File: src/models/connect-four-games.ts [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+	}
+
+	/** Randomly pick which player moves first and start the game */
+	public async acceptExistingChallengeAndStartGame(existingChallenge: IConnectFour) {
+		const { _id, red_user_id, yellow_user_id } = existingChallenge;
+		const to_move = Math.random() < 0.5 ? red_user_id : yellow_user_id;
+		return await super.updateOne({ _id }, { to_move, is_accepted: true });
```
Comment (judegiordano) [inline src/models/connect-four-games.ts]: dont await here, return the unresolved promise

File: src/models/connect-four-games.ts [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+		)
+			return existingChallenge;
+		const { _id } = existingChallenge;
+		return await super.updateOne({ _id }, { yellow_user_id: challengedUser.user_id, wager });
+	}
+
+	/** Randomly pick which player moves first and start the game */
+	public async acceptExistingChallengeAndStartGame(existingChallenge: IConnectFour) {
+		const { _id, red_user_id, yellow_user_id } = existingChallenge;
+		const to_move = Math.random() < 0.5 ? red_user_id : yellow_user_id;
```
Comment (judegiordano) [inline src/models/connect-four-games.ts]: i think i added the chance library for seeding random values somewhere in lib or helpers

File: src/models/connect-four-games.ts [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+					board[i + 1][j - 1] == color &&
+					board[i + 2][j - 2] == color &&
+					board[i + 3][j - 3] == color
+				)
+					return true;
+			}
+		}
+
+		return false;
+	}
```
Comment (judegiordano) [inline src/models/connect-four-games.ts:221]: my god

File: src/models/connect-four-games.ts [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+		for (const column of game.board) {
+			if (column.length < 6) return false;
+		}
+		return true;
+	}
+
+	/** Update the gamestate at the end of a turn */
+	public async endTurn(game: IConnectFour) {
+		const { _id } = game;
+		return await super.updateOne({ _id }, game);
```
Comment (judegiordano) [inline src/models/connect-four-games.ts]: dont await here, return the unresolved promise

File: src/controllers/gambling/index.ts (hunk 1 above)
Comment (cdleveille) [reply src/controllers/gambling/index.ts:278]: Refactored this whole section, should be better now.

**Review by judegiordano — APPROVED** [review]

**Pull request merged**